### PR TITLE
WL21484 - Add an additional parameter for entity animations to "allow translation"

### PR DESCRIPTION
**Pull request by Menithal**
This PR handles [WL21484](https://worklist.net/21484).

 ### Protocol Update

The enabled content creators to disable / enable translations on Model Entity Animations. This allows many animations to be used on models which have the same skeleton naming, but different proportions of bones.

This basically allows the creations of NPC's using Entities, instead of AC run ones.

### Test Plan:

1. Run the following in the Script console
``` 
var m = "http://norteclabs.com/HF/experiment/PopupTest.fbx";

Entities.addEntity(
{ type: "Model",
  name: "Tester",
  position: Vec3.sum(MyAvatar.position, Vec3.multiplyQbyV(MyAvatar.orientation, {x:0,y:0,z:-2})), 
  modelURL: m,
  animation:{url:m, running: true}});

Entities.addEntity(
{ type: "Model",
  name: "Control",
  position: Vec3.sum(MyAvatar.position, Vec3.multiplyQbyV(MyAvatar.orientation, {x:0,y:0,z:-4})), 
  modelURL: m,
  animation:{url:m, running: true}});
``` 

#### Basic

1. This should create two model entities that consist of two (inverted) boxes, with the top one with a spinning back and forth, and **bobs up and down**.  This tests that by default, the Allow Translation is active if not defined.
2.  Selecting the closest (*Tester*) with CREATE should show a new field under Animation, which is "Animation Allow Translation" which should be enabled. (by default to not break existing content)
3. Disable "Animation Allow Translation" for "Tester". 
4. *Tester*'s top box should **Stop Bobbing up and down, but keep spinning**.  Control should not be affected.
5. Re-enable "Animation Allow Translation"  for "Tester"
6. *Tester*'s top box should **Start bobbing up and down and keep spinning**. Control should not change.

### Export / Import

7. Disable "Animation Allow Translation" for *Tester*
8. Export both *Tester* and *Control* together into to JSON.
9. Delete *Tester* and *Control* from the World
10. Import both *Tester* and *Control* from the JSON
11. *Tester* should still have allowTranslation false, 
12. *Control* should have allowTranslation true.

Above tests import and export steps export the entity date correctly

### Finalization

13. Quit High Fidelity
14. Open High Fidelity again and connect to the same domain the test object are at
15. *Tester* should still have allowTranslation false, 
16. *Control* should have allowTranslation true.

Above tests make sure domain entity settings are consistent.

### CODE REVIEW

**Comment by Menithal** [comment]
Ugh. There specifically is a big refactor in RenderableModelEntityItem that makes quick rebase or merge impossible and requires me to redo the logic I had planned out in this build. So Ill have to throw that out of the window and re-do the changes.

DO NOT Test until this has been addressed.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/3014/HighFidelity-Beta-PR11290-5506520226e65025898348b9a0198a09d91dd619.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/3014/HighFidelity-Beta-PR11290-5506520226e65025898348b9a0198a09d91dd619.exe

**Comment by Menithal** [comment]
# Ready To Test.

**Comment by YuppyPlays** [comment]
Testing.

**Comment by YuppyPlays** [comment]
Testing complete:

Run the following in the Script console **Check**
This should create two model entities that consist of two (inverted) boxes, with the top one with a spinning back and forth, and bobs up and down. This tests that by default, the Allow Translation is active if not defined. **Check**
Selecting the closest ("Tester") with CREATE should show a new field under Animation, which is "Animation Allow Translation" which should be enabled. (by default to not break existing content) **Check**
Disable "Animation Allow Translation" for "Tester". **Check**
"Tester"'s top box should Stop Bobbing up and down, but keep spinning. Control should not be affected. **Check**
Re-enable "Animation Allow Translation" for "Tester" **Check**
"Tester"'s top box should Start bobbing up and down and keep spinning. Control should not change. **Check**

**Pass**

**Comment by Menithal** [comment]
Thanks @YuppyPlays , I had forgotten to add some more steps to the test.

I added steps 7-13, could you test those as well?

**Comment by Nex-Pro** [comment]
Testing this.

**Comment by Nex-Pro** [comment]
## PASS

***Also confimed that inverted face normals were due to negative scale of entity prior to exporting the FBX from the 3d app.***

**Loaded some other animated FBX files to check they still work.**

This should create two model entities that consist of two (inverted) boxes, with the top one with a spinning back and forth, and bobs up and down. This tests that by default, the Allow Translation is active if not defined. **DONE**

Selecting the closest (Tester) with CREATE should show a new field under Animation, which is "Animation Allow Translation" which should be enabled. (by default to not break existing content) **PASS**

Disable "Animation Allow Translation" for "Tester". **DONE**

Tester's top box should Stop Bobbing up and down, but keep spinning. **SUCCESS**
Control should not be affected. **SUCCESS**
Re-enable "Animation Allow Translation" for "Tester" **DONE**

Tester's top box should Start bobbing up and down and keep spinning. **SUCCESS**
Control should not change. **SUCCESS**

### Export / Import

Disable "Animation Allow Translation" for Tester **DONE**
Export both Tester and Control together into to JSON. **DONE**
Delete Tester and Control from the World **DONE**
Import both Tester and Control from the JSON **DONE**
Tester should still have allowTranslation false, **SUCCESS**
Control should have allowTranslation true. **SUCCESS**
Above tests import and export steps export the entity data correctly

Finalization

Quit High Fidelity
Open High Fidelity again and connect to the same domain the test object are at
Tester should still have allowTranslation false, **SUCCESS**
Control should have allowTranslation true. **SUCCESS**

@Menithal  I feel this needs a full smoke test.

**Comment by Menithal** [comment]
@Nex-Pro Agreed.

File: libraries/entities/src/EntityPropertyFlags.h [hunk 1]
```diff
@@ -40,6 +40,7 @@ enum EntityPropertyList {
     PROP_ANIMATION_FPS,
     PROP_ANIMATION_FRAME_INDEX,
     PROP_ANIMATION_PLAYING,
+    PROP_ANIMATION_ALLOW_TRANSLATION,
```
Comment (sethalves) [inline libraries/entities/src/EntityPropertyFlags.h]: is this safe?  the comment below says to add new properties lower down.

Comment (sethalves) [reply libraries/entities/src/EntityPropertyFlags.h]: apparently this is okay.

**Comment by alphaversiond** [comment]
This failed my testing. Could you please look into this further? I will provide FBX to test with.

small

bone_translation_failed

**Comment by alphaversiond** [comment]
The model on the far left is one that uses pre-existing situation. Meaning, translations are ON. The others, translations are OFF.

**Pull request merged**